tools/dataset.cpp

```diff
@@ -1 +1,199 @@
+    status = parquet::arrow::OpenFile(input, pool, &arrow_reader);
+    return_if_error(status.ok(), c.error, 0, "Can't open file");
+
+    // Read File into table
+    std::shared_ptr<arrow::Table> table;
+    status = arrow_reader->ReadTable(&table);
+    return_if_error(status.ok(), c.error, 0, "Can't read file");
+
+    fill_array(c, max_batch_size, table);
+}
+
+void import_csv(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    arrow::io::IOContext io_context = arrow::io::default_io_context();
+    auto maybe_input = arrow::io::ReadableFile::Open(c.paths_pattern);
+    return_if_error(maybe_input.ok(), c.error, 0, "Can't open file");
+    std::shared_ptr<arrow::io::InputStream> input = *maybe_input;
+
+    auto read_options = arrow::csv::ReadOptions::Defaults();
+    auto parse_options = arrow::csv::ParseOptions::Defaults();
+    auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+    // Instantiate TableReader from input stream and options
+    auto maybe_reader = arrow::csv::TableReader::Make(io_context, input, read_options, parse_options, convert_options);
+    return_if_error(maybe_reader.ok(), c.error, 0, "Can't instatinate reader");
+    std::shared_ptr<arrow::csv::TableReader> reader = *maybe_reader;
+
+    // Read table from CSV file
+    auto maybe_table = reader->Read();
+    return_if_error(maybe_table.ok(), c.error, 0, "Can't read file");
+    std::shared_ptr<arrow::Table> table = *maybe_table;
+
+    fill_array(c, max_batch_size, table);
+}
+
+////////// Parsing with SIMDJSON //////////
+
+simdjson::ondemand::document& rewinded(simdjson::ondemand::document& doc) noexcept {
+    doc.rewind();
+    return doc;
+}
+
+simdjson::ondemand::object& rewinded(simdjson::ondemand::object& doc) noexcept {
+    doc.reset();
+    return doc;
+}
+
+void import_json(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    std::vector<edge_t> array;
+    bool edge_state = (c.edge_id_field != "edge");
+
+    auto handle = open(c.paths_pattern, O_RDONLY);
+    return_if_error(handle != -1, c.error, 0, "Can't open file");
+
+    auto begin = mmap(NULL, c.paths_length, PROT_READ, MAP_PRIVATE, handle, 0);
+    std::string_view mapped_content = std::string_view(reinterpret_cast<char const*>(begin), c.paths_length);
+    madvise(begin, c.paths_length, MADV_SEQUENTIAL);
+
+    // https://github.com/simdjson/simdjson/blob/master/doc/basics.md#newline-delimited-json-ndjson-and-json-lines
+    simdjson::ondemand::parser parser;
+    simdjson::ondemand::document_stream docs = parser.iterate_many( //
+        mapped_content.data(),
+        mapped_content.size(),
+        1000000ul);
+
+    ukv_key_t source = 0;
+    ukv_key_t target = 0;
+    ukv_key_t edge = 0;
+
+    for (auto doc : docs) {
+        simdjson::ondemand::object data = doc.get_object().value();
+        source = rewinded(data)[c.source_id_field];
+        target = rewinded(data)[c.target_id_field];
+        edge = edge_state ? rewinded(data)[c.edge_id_field] : 0;
```

Comment:
edge ids have default value, check constants in the header

Reply:
Fixed